osmoutils/store_helper.go

```diff
@@ -222,3 +225,66 @@ func DeleteAllKeysFromPrefix(store store.KVStore, prefixKey []byte) {
 		prefixStore.Delete(iter.Key())
 	}
 }
+
+// GetCoinArrayFromPrefix returns all coins from the store that has the given prefix.
+func GetCoinArrayFromPrefix(ctx sdk.Context, storeKey storetypes.StoreKey, storePrefix []byte) []sdk.Coin {
+	coinArray := make([]sdk.Coin, 0)
+
+	store := ctx.KVStore(storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, storePrefix)
+
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		bz := iterator.Value()
+		sdkInt := osmomath.Int{}
+		if err := sdkInt.Unmarshal(bz); err == nil {
+			// Split the key into prefix and denom
+			keyParts := strings.Split(string(iterator.Key()), "|")
+			if len(keyParts) > 1 {
+				denom := keyParts[1]
+				coinArray = append(coinArray, sdk.NewCoin(denom, sdkInt))
+			}
+		}
+	}
+
+	return coinArray
+}
+
+// GetCoinByDenomFromPrefix returns the coin from the store that has the given prefix and denom.
+// If the denom is not found, a zero coin is returned.
+func GetCoinByDenomFromPrefix(ctx sdk.Context, storeKey storetypes.StoreKey, storePrefix []byte, denom string) (sdk.Coin, error) {
+	store := prefix.NewStore(ctx.KVStore(storeKey), storePrefix)
+	key := []byte(fmt.Sprintf("%s%s%s", storePrefix, "|", []byte(denom)))
+
+	bz := store.Get(key)
+	if len(bz) == 0 {
+		return sdk.NewCoin(denom, osmomath.ZeroInt()), nil
+	}
+
+	sdkInt := osmomath.Int{}
+	if err := sdkInt.Unmarshal(bz); err != nil {
+		return sdk.NewCoin(denom, osmomath.ZeroInt()), err
+	}
+
+	return sdk.NewCoin(denom, sdkInt), nil
+}
+
+// IncreaseCoinByDenomFromPrefix increases the coin from the store that has the given prefix and denom by the specified amount.
+func IncreaseCoinByDenomFromPrefix(ctx sdk.Context, storeKey storetypes.StoreKey, storePrefix []byte, denom string, increasedAmt osmomath.Int) error {
+	store := prefix.NewStore(ctx.KVStore(storeKey), storePrefix)
+	key := []byte(fmt.Sprintf("%s%s%s", storePrefix, "|", []byte(denom)))
```

Comment:
were double prefixing by the store prefix, resulting in the odd behavior

Reply:
https://github.com/osmosis-labs/osmosis/pull/7240/commits/b83740e0cc91412877cb190fe5efa92d2c6a3b7f

Reply:
That LGTM